doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -206,90 +165,198 @@ simply :g:`t=u` dropping the implicit type of :g:`t` and :g:`u`.
-      :name: stepr
+   .. tacn:: stepr @one_term {? by @ltac_expr }
 
-      This behaves as :tacn:`stepl` but on the right-hand-side of the binary
-      relation. Lemmas are expected to be of the form
+      This behaves like :tacn:`stepl` but on the right hand side of the binary
+      relation. Lemmas are expected to be in the form
       :g:`forall x y z, R x y -> eq y z -> R x z`.
 
-   .. cmd:: Declare Right Step @term
+   .. cmd:: Declare Right Step @one_term
 
        Adds :n:`@term` to the database used by :tacn:`stepr`.
 
+Performance-oriented tactic variants
+------------------------------------
 
-.. tacn:: change @term
-   :name: change
+.. tacn:: change @conversion {? @occurrences }
 
-   This tactic applies to any goal. It implements the rule ``Conv`` given in
+   .. insertprodn conversion conversion
+
+   .. prodn::
+      conversion ::= @one_term {? {? at @occs_nums } with @one_term }
+
+   .. todo PR: need explanation of conversion in a couple tactics, should share
+      no idea what the "with" clause does
+
+   Implements the rule ``Conv`` given in
    :ref:`subtyping-rules`. :g:`change U` replaces the current goal `T`
-   with `U` providing that `U` is well-formed and that `T` and `U` are
+   with `U` provided that `U` is well-formed and that `T` and `U` are
    convertible.
 
+   :n:`{? at @occs_nums } with @one_term__repl`
+     Replaces the occurrences specified by :n:`@occs_nums` of :n:`@one_term`
+     with :n:`@one_term__repl`.
+
    .. exn:: Not convertible.
       :undocumented:
 
-   .. tacv:: change @term with @term’
-
-      This replaces the occurrences of :n:`@term` by :n:`@term’` in the current goal.
-      The term :n:`@term` and :n:`@term’` must be convertible.
-
-   .. tacv:: change @term at {+ @natural} with @term’
-
-      This replaces the occurrences numbered :n:`{+ @natural}` of :n:`@term` by :n:`@term’`
-      in the current goal. The terms :n:`@term` and :n:`@term’` must be convertible.
+   .. exn:: Too few occurrences.
+      :undocumented:
 
-      .. exn:: Too few occurrences.
-         :undocumented:
+   .. todo PR: Need to integrate the following elsewhere.  But how are
+      the occs_nums used?  Should the syntax be improved, seems a little confusing.
+      Maybe ask @herbelin?
 
-   .. tacv:: change @term {? {? at {+ @natural}} with @term} in @goal_occurrences
+   .. tacv:: change @one_term {? {? at {+ @occs_nums } } with @one_term__repl } {? @occurrences }
 
       In the presence of :n:`with`, this applies :tacn:`change` to the
-      occurrences specified by :n:`@goal_occurrences`. In the
-      absence of :n:`with`, :n:`@goal_occurrences` is expected to
+      occurrences specified by :n:`@occurrences`. In the
+      absence of :n:`with`, :n:`@occurrences` is expected to
       only list hypotheses (and optionally the conclusion) without
       specifying occurrences (i.e. no :n:`at` clause).
 
-   .. tacv:: now_show @term
+   .. tacn:: now_show @conversion
```

Comment:
Actually `now_show` can only take a `one_term` as argument.

Reply:
Why do you say that?  I see this:

```
| "change" conversion clause_dft_concl

Ltac now_show c := change c.
```

Reply:
I think your misunderstanding what `Ltac` does. It doesn't pass everything it gets syntactically, this `Ltac` takes a single argument and thus doesn't use the `with` clause of `change`.

Reply:
Ah, OK.  Easy to get confused since Ltac semantics are vague, also too many ways to define tactics, all with different capabilities and quirks.  And constantly switching between them all.